File: client/cody-shared/src/chat/transcript/interaction.ts
import type { ContextFile, ContextMessage, PreciseContext } from '../../codebase-context/messages'

import type { ChatMessage, InteractionMessage } from './messages'

export interface InteractionJSON {
    humanMessage: InteractionMessage
    assistantMessage: InteractionMessage
    fullContext: ContextMessage[]
    usedContextFiles: ContextFile[]
    usedPreciseContext: PreciseContext[]
    timestamp: string

    // DEPRECATED: Legacy field for backcompat, renamed to `fullContext`
    context?: ContextMessage[]
}

export class Interaction {
    constructor(
        private readonly humanMessage: InteractionMessage,
        private assistantMessage: InteractionMessage,
        private fullContext: Promise<ContextMessage[]>,
        private usedContextFiles: ContextFile[],
        private usedPreciseContext: PreciseContext[] = [],
        public readonly timestamp: string = new Date().toISOString()
    ) {}

    public getAssistantMessage(): InteractionMessage {
        return { ...this.assistantMessage }
    }

    public setAssistantMessage(assistantMessage: InteractionMessage): void {
        this.assistantMessage = assistantMessage
    }

    public getHumanMessage(): InteractionMessage {
        return { ...this.humanMessage }
    }

    public async getFullContext(): Promise<ContextMessage[]> {
        const msgs = await this.fullContext
        return msgs.map(msg => ({ ...msg }))
    }

    public async hasContext(): Promise<boolean> {
        const contextMessages = await this.fullContext
        return contextMessages.length > 0
    }

    public setUsedContext(usedContextFiles: ContextFile[], usedPreciseContext: PreciseContext[]): void {
        this.usedContextFiles = usedContextFiles
        this.usedPreciseContext = usedPreciseContext
    }

    /**
     * Converts the interaction to chat message pair: one message from a human, one from an assistant.
     */
    public toChat(): ChatMessage[] {
        return [
            this.humanMessage,
            {
                ...this.assistantMessage,
                contextFiles: this.usedContextFiles,
                preciseContext: this.usedPreciseContext,
            },
        ]
    }

    public async toChatPromise(): Promise<ChatMessage[]> {
        await this.fullContext
        return this.toChat()
    }

    public async toJSON(): Promise<InteractionJSON> {
        return {
            humanMessage: this.humanMessage,
            assistantMessage: this.assistantMessage,
            fullContext: await this.fullContext,
            usedContextFiles: this.usedContextFiles,
            usedPreciseContext: this.usedPreciseContext,
            timestamp: this.timestamp,
        }
    }
}

File: client/jetbrains/src/main/java/com/sourcegraph/cody/chat/Interaction.java
package com.sourcegraph.cody.chat;

import com.sourcegraph.cody.context.ContextMessage;
import java.util.List;
import java.util.Optional;

public class Interaction {
  private final ChatMessage humanMessage;
  private final List<ContextMessage> fullContext;
  private ChatMessage assistantMessage;

  public Interaction(ChatMessage humanMessage, List<ContextMessage> fullContext) {
    this.humanMessage = humanMessage;
    this.fullContext = fullContext;
    this.assistantMessage = ChatMessage.createAssistantMessage("");
  }

  public void setAssistantMessage(ChatMessage assistantMessage) {
    this.assistantMessage = assistantMessage;
  }

  public boolean hasContext() {
    return !this.fullContext.isEmpty();
  }

  public Optional<ChatMessage> getAssistantMessage() {
    return Optional.ofNullable(assistantMessage);
  }

  public ChatMessage getHumanMessage() {
    return humanMessage;
  }

  public List<ContextMessage> getFullContext() {
    return fullContext;
  }
}
